### PR TITLE
Adds the ability to use environment variables for connection

### PR DESCRIPTION
**Pull request by neovintage**
We store most everything in env variables for database connections. The intention is if a env variable is set for the connection that should take priority. Would love some feedback on this.

### CODE REVIEW

**Comment by neovintage** [comment]
@mistercrunch any interest in this pull request? I'm trying to get this working on Heroku so we can use airflow at Heroku. :)

**Comment by mistercrunch** [comment]
Oh yes oh yes! I'm just a bit behind on code review. Reviewing today

**Comment by mistercrunch** [comment]
Thank you for the PR btw!

**Comment by mistercrunch** [comment]
So the reason why it took me a moment to get to this PR is that I was a bit confused.

Are you trying to create your Connection objects from env vars, or do you want to bypass the Connection model altogether? Because from reading the code it doesn't see like it would...

**Comment by artwr** [comment]
It seems to be about creating a Connection from an env var. Would it be possible to separate specifying the connection params through a URI and an env var? I could see the URI specification to be useful, even if not provided through an environment variable.

**Comment by mistercrunch** [comment]
Hey if what you are trying to do is bypass storing the `Connection` in the metadata database altogether, you might be able to hijack to `get_connection` method of `BaseHook`. 
https://github.com/airbnb/airflow/blob/master/airflow/hooks/base_hook.py#L35

You could look for the existence an env var `PREFIX + conn_id` and if it exists you parse it and return a Connection object without it ever hitting the database. 

That'd be a pretty sweet feature as right now the connection info is stored in plain in the metadata database and many people may want to have another approach.

**Comment by neovintage** [comment]
@artwr is right. Its about creating a connection from an env var. 

@mistercrunch I don't necessarily want to bypass the metadata database. Although, would it be worthwhile to do both? Meaning, we could have the `PREFIX + conn_id` method and save an env var associated to a conn_id in the metadata database. I like the idea of being more explicit about what what `conn_id` means when I'm working in my dags meaning I should save the env var in the metadata db. 

As a side note, I also have another branch sitting in my repo with some crypto that will prevent passwords from being stored in plain text. I'll have to get that one cleaned up.

**Comment by mistercrunch** [comment]
Ok, now that I understand what you are up to, why not just doing it all in the `__init__` method , as in:

```
# warning, heavily simplified semi-pseudocode
class Connection(Base):
     def __init__(...., env_var):
         if env_var:
             self.from_env_var(env_var)
         else:
             self.host = host
             .....

     def from_env_var(self, env_var):

         parsed_uri = sqla.parse_uri(os.environ[env_var])
         self.host = parsed_uri.host
         ...
```

I don't think there's a need for getters setters. Welcome to python! :)

**Comment by neovintage** [comment]
ha. nice warning in the comments. Give me a few moments to rework this.

**Comment by danielzohar** [comment]
:+1: for this PR!

**Comment by neovintage** [comment]
I worked through this again and tried to keep the scope small. I updated the base_hook to first see if the conn_id is an environment variable before checking the connections in the database. This way the connection information isn't saved in the database. The connection model has been refactored to parse a uri and then assign it to all the appropriate properties. 

@mistercrunch let me know what you think.

File: airflow/hooks/base_hook.py [hunk 1]
```diff
@@ -1,5 +1,6 @@
 from builtins import object
 import logging
+import os
 import random
```
Comment (mistercrunch) [inline airflow/hooks/base_hook.py:5]: Let's use a prefix for env vars as a module constant:

```
CONN_ENV_PREFIX = 'AIRFLOW_CONN_'
```

File: airflow/hooks/base_hook.py [hunk 2]
```diff
@@ -33,7 +34,11 @@ def get_connections(self, conn_id):
         return db
 
     def get_connection(self, conn_id):
```
Comment (mistercrunch) [inline airflow/hooks/base_hook.py]: Let's call os.environ.get only once:
`env_conn_uri = os.environ.get(CONN_ENV_PREFIX  + conn_id)`

File: airflow/hooks/base_hook.py [hunk 3]
```diff
@@ -33,7 +34,11 @@ def get_connections(self, conn_id):
...
-        conn = random.choice(self.get_connections(conn_id))
+        if os.environ.get(conn_id):
+            temp_uri = urlparse(os.environ.get(conn_id))
+            conn = Connection(uri=temp_uri)
+        if conn is None:
```
Comment (mistercrunch) [inline airflow/hooks/base_hook.py]: "Referenced before assignement"?

File: airflow/models.py [hunk 4]
```diff
@@ -311,22 +312,37 @@ class Connection(Base):
...
+            self.login = login
+            self.password = password
+            self.schema = schema
+            self.port = port
+            self.extra = extra
+
+    def parse_from_uri(self, uri):
+        temp_uri = urlparse(uri)
+        hostname = temp_uri.hostname or ''
+        if '%2f' in hostname:
```
Comment (mistercrunch) [inline airflow/models.py:339]: I'm surprised that `urlparse` doens't urldecode... I guess hostname is not expected to have slashes in it...

**Comment by mistercrunch** [comment]
It'd be nice to add a note here in the docs:
http://pythonhosted.org/airflow/installation.html#connections (maps to this file: https://github.com/airbnb/airflow/blob/master/docs/installation.rst)

And a quick note here too:
https://github.com/airbnb/airflow/blob/master/docs/concepts.rst#connections

**Comment by neovintage** [comment]
Updates based on comments.

**Comment by danielepolencic** [comment]
@neovintage any luck with airflow on Heroku? I'd like to do the same, but I'm not well versed in python. Could you share how you did it?

**Comment by neovintage** [comment]
@danielepolencic Yes. I've got it working. The reference project to get it running is here: https://github.com/heroku/heroku-airflow

Unfortunately, I haven't had time to document it and how to get started. The process is a little contrived because Heroku Button doesn't work for our new docker based workflow (https://devcenter.heroku.com/articles/docker) which makes it a little hard to get started.

**Pull request merged**